packages/go_router/lib/src/delegate.dart

```diff
@@ -44,13 +44,21 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
 
   final GlobalKey<NavigatorState> _key = GlobalKey<NavigatorState>();
 
+  /// The list of completers for the promises when pushing asynchronous routes.
+  final Map<String, Completer<dynamic>> _completers =
+      <String, Completer<dynamic>>{};
   RouteMatchList _matches = RouteMatchList.empty();
   final Map<String, int> _pushCounts = <String, int>{};
 
-  /// Pushes the given location onto the page stack
-  void push(RouteMatch match) {
-    // Remap the pageKey to allow any number of the same page on the stack
+  /// Pushes the given location onto the page stack with an optional promise.
+  Future<T?> push<T extends Object?>(RouteMatch match) {
+    // Remap the pageKey to allow any number of the same page on the stack.
     final String fullPath = match.fullpath;
+
+    // Create a completer for the promise and store it in the completers map.
+    final Completer<T> completer = Completer<T>();
+    _completers[fullPath] = completer;
```

Comment:
It's should be replaced by `final Completer<T?> completer = Completer<T?>();`
If T is a nonnull type, app will crash when users pressing the Back button (complete with a null value).

Reply:
It will not because T extends Object?, so T is out of the box nullable

Reply:
If I write:
`Future<bool?> result = context.push("someRoute")`
T will be inferred as `bool`, not `bool?`. I can only fix by specifying T:
`Future<bool?> result = context.push<bool?>("someRoute")`

I tested with flutter master channel.

Reply:
I changed it, hope that fixes it